dll/win32/kernel32/kernel32_vista/vista.c

```diff
@@ -26,61 +26,61 @@
  */
 BOOL
 WINAPI
-QueryFullProcessImageNameW(HANDLE hProcess,
-                           DWORD dwFlags,
-                           LPWSTR lpExeName,
-                           PDWORD pdwSize)
+QueryFullProcessImageNameW(
+    _In_ HANDLE hProcess,
+    _In_ DWORD dwFlags,
+    _Out_writes_to_(*lpdwSize, *lpdwSize) LPWSTR lpExeName,
+    _Inout_ PDWORD lpdwSize)
 {
-    BYTE Buffer[sizeof(UNICODE_STRING) + MAX_PATH * sizeof(WCHAR)];
-    UNICODE_STRING *DynamicBuffer = NULL;
-    UNICODE_STRING *Result = NULL;
+    BOOL bRet;
+    DWORD dwBufferSize;
+    PUNICODE_STRING pBuffer;
     NTSTATUS Status;
-    DWORD Needed;
+    USHORT uCch;
 
-    Status = NtQueryInformationProcess(hProcess,
-                                       ProcessImageFileName,
-                                       Buffer,
-                                       sizeof(Buffer) - sizeof(WCHAR),
-                                       &Needed);
-    if (Status == STATUS_INFO_LENGTH_MISMATCH)
+    if (dwFlags && dwFlags != PROCESS_NAME_NATIVE)
     {
-        DynamicBuffer = RtlAllocateHeap(RtlGetProcessHeap(), 0, Needed + sizeof(WCHAR));
-        if (!DynamicBuffer)
-        {
-            BaseSetLastNTError(STATUS_NO_MEMORY);
-            return FALSE;
-        }
-
-        Status = NtQueryInformationProcess(hProcess,
-                                           ProcessImageFileName,
-                                           (LPBYTE)DynamicBuffer,
-                                           Needed,
-                                           &Needed);
-        Result = DynamicBuffer;
+        BaseSetLastNTError(STATUS_INVALID_PARAMETER_2);
+        return FALSE;
     }
-    else Result = (PUNICODE_STRING)Buffer;
-
-    if (!NT_SUCCESS(Status)) goto Cleanup;
 
-    if (Result->Length / sizeof(WCHAR) + 1 > *pdwSize)
+    dwBufferSize = *lpdwSize * sizeof(WCHAR) + sizeof(UNICODE_STRING);
+    pBuffer = (PUNICODE_STRING)RtlAllocateHeap(RtlGetProcessHeap(), 0, dwBufferSize);
+    if (!pBuffer)
     {
-        Status = STATUS_BUFFER_TOO_SMALL;
-        goto Cleanup;
+        BaseSetLastNTError(STATUS_NO_MEMORY);
+        return FALSE;
     }
 
-    *pdwSize = Result->Length / sizeof(WCHAR);
-    memcpy(lpExeName, Result->Buffer, Result->Length);
-    lpExeName[*pdwSize] = 0;
-
-Cleanup:
-    RtlFreeHeap(RtlGetProcessHeap(), 0, DynamicBuffer);
-
-    if (!NT_SUCCESS(Status))
+    bRet = FALSE;
+    Status = NtQueryInformationProcess(hProcess,
+                                       dwFlags ? ProcessImageFileName : ProcessImageFileNameWin32,
```

Comment:
```suggestion
                                       dwFlags == PROCESS_NAME_NATIVE ? ProcessImageFileName : ProcessImageFileNameWin32,
```

Reply:
What if a future update adds another flag? Then this code would break. The proper way of doing it, without relying on assumptions is `(dwFlags & PROCESS_NAME_NATIVE) ? ProcessImageFileName : ProcessImageFileNameWin32,`

Reply:
It is called `dwFlags` for a reason. So far only one flag exists. And if MS ever adds a flag, then your code becomes invalid. Why do you insist on keeping fragile code?